src/python/aqueduct_executor/operators/connectors/tabular/s3.py

```diff
@@ -29,58 +36,121 @@ def authenticate(self) -> None:
 
-    def _fetch_object(self, key: str, format: common.S3FileFormat) -> pd.DataFrame:
+    def _fetch_object(self, key: str, params: extract.S3Params) -> Any:
         response = self.s3.Object(self.bucket, key).get()
         data = response["Body"].read()
-        buf = io.BytesIO(data)
-        if format == common.S3FileFormat.CSV:
-            return pd.read_csv(buf)
-        elif format == common.S3FileFormat.JSON:
-            return pd.read_json(buf)
-        elif format == common.S3FileFormat.PARQUET:
-            return pd.read_parquet(buf)
+        if params.artifact_type == ArtifactType.TABULAR:
+            if params.format is None:
+                raise Exception("You must specify a file format for tabular data.")
+            buf = io.BytesIO(data)
+            if params.format == common.S3TabularFormat.CSV:
+                return pd.read_csv(buf)
+            elif params.format == common.S3TabularFormat.JSON:
+                return pd.read_json(buf)
+            elif params.format == common.S3TabularFormat.PARQUET:
+                return pd.read_parquet(buf)
+            raise Exception("Unknown S3 file format %s." % params.format)
+        elif params.artifact_type == ArtifactType.JSON:
+            # This assumes that the encoding is "utf-8". May worth considering letting the user
+            # specify custom encoding in the future.
+            json_data = data.decode(_DEFAULT_JSON_ENCODING)
+            # Make sure the data is a valid json object.
+            json.loads(json_data)
+            return json_data
+        elif params.artifact_type == ArtifactType.IMAGE:
+            return Image.open(io.BytesIO(data))
+        elif params.artifact_type == ArtifactType.BYTES:
+            return data
+        elif (params.artifact_type == ArtifactType.STRING or
+              params.artifact_type == ArtifactType.BOOL or
+              params.artifact_type == ArtifactType.NUMERIC or
+              params.artifact_type == ArtifactType.DICT or
+              params.artifact_type == ArtifactType.TUPLE or
+              params.artifact_type == ArtifactType.PICKLABLE):
+            unpickled_data = pickle.loads(data)
 
-        raise Exception("Unknown S3 file format %s." % format)
+            if params.artifact_type == ArtifactType.STRING:
+                assert(isinstance(unpickled_data, str))
+            elif params.artifact_type == ArtifactType.BOOL:
+                assert(isinstance(unpickled_data, bool) or isinstance(unpickled_data, np.bool_))
+            elif params.artifact_type == ArtifactType.NUMERIC:
+                assert(isinstance(unpickled_data, int) or isinstance(unpickled_data, float) or isinstance(unpickled_data, np.number))
+            elif params.artifact_type == ArtifactType.DICT:
+                assert(isinstance(unpickled_data, dict))
+            elif params.artifact_type == ArtifactType.TUPLE:
+                assert(isinstance(unpickled_data, tuple))
```

Comment:
I don't think these should be assertions, since they rely on user inputs, so they're not validating internal consistency without our system. Can we raise an exception with a useful error message?

Reply:
Good catch, will fix!